man/gcmorse.1
.TH gcmorse 1 "10th March 2019" "v1.1\-1"
.SH NAME
.PP
gcmorse \- text to Morse code converter (and vice versa)
.SH SYNOPSIS
.PP
gcmorse [\fB\-a\fP] [\fB\-d\fP] [\fB\-h\fP] [\fB\-i\fP \fIfilename\fP] [\fB\-o\fP \fIfilename\fP] [\fB\-p\fP] [\fB\-u\fP] [\fB\-v\fP]
.SH DESCRIPTION
.PP
Converts Morse code to text, or vice versa. Currently, only ASCII characters are supported.
If input file is not specified, gcmorse will read stdin, until EOF, or EOT character occurs.
Similarly, if output file is not specified, gcmorse will write output to stdout.
If \fB\-a\fP option is not set, gcmorse is going to \fBoverwrite output file content\fP!
.SS Morse code format
.PP
Morse code is written as lines, dots and spaces.
There is one space between each two letter in a word, and words are separated by two spaces.
Normally, all characters are outputted lowercase (unless \-p option is used).
Also, gcmorse does not guarantee that line spacing will be exactly the same after encoding and decoding.
.SH OPTIONS
.PP
\fB\-a\fP, \fB\-\-append\fP \- append data to file instead of overwriting it
.PP
\fB\-d\fP, \fB\-\-decode\fP \- decode from Morse code to text file (assumes input file is in Morse code format)
.PP
\fB\-h\fP, \fB\-\-help\fP \- display a help message
.PP
\fB\-i\fP, \fB\-\-input\fP \fIfilename\fP \- specifies input file name. Given file should not contain any non\-ASCII characters, otherwise they won't be encrypted and gcmorse will throw errors.
.PP
\fB\-o\fP, \fB\-\-output\fP \fIfilename\fP \- specifies output file name
.PP
\fB\-p\fP, \fB\-\-noprosigns\fP \- do not change special characters (new line, etc.) to Morse prosigns (other Morse code applications may not handle them way gcmorse does)
.PP
\fB\-u\fP, \fB\-\-uppercase\fP \- output uppercase letters when decoding
.PP
\fB\-v\fP, \fB\-\-version\fP \- display a version information
.SH EXAMPLES
.PP
\fIRecommended way of writing to files:\fP
    gcmorse \-i testfile2.txt \-o output2.morse
    gcmorse \-d \-i output2.morse \-o output2.txt
.PP
\fIThe second way:\fP
    gcmorse \-i testfile.txt > output.morse
    gcmorse \-d \-i output.morse > output.txt
.SH BUGS
.PP
There are no known bugs yet, if you find some, please report them here:
\[la]https://github.com/Jacajack/cmorse/issues\[ra]
.SH AUTHOR
.PP
Bruno Ramalhete \[la]bram.512@gmail.com\[ra]
.PP
Inspired by cmorse application by Jacek Wieczorek
.SH COPYRIGHT
.PP
Copyright (c) 2019 Bruno Ramalhete
.PP
The GNU General Public License v3.0


src/gcmorse.c
/*
	gcmorse - simple text to Morse code converter
	Copyright (C) 2019	Bruno Ramalhete <bram.512@gmail.com>

	This file is part of gcmorse.

	GCmorse is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	GCmorse is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "gcmorse.h"

//Display version number
void version( int exitcode )
{
	fprintf( stderr, "gcmorse " VERSION "\n\n\r" );
	fprintf( stderr,
		"This program is distributed in the hope that it will be useful,\n\r"
		"but WITHOUT ANY WARRANTY; without even the implied warranty of\n\r"
		"MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n\r"
		"GNU General Public License for more details.\n\r" );
	exit( exitcode );
}

//Display help message
void help( int exitcode )
{
	fprintf( stderr, "gcmorse " VERSION "\n\n\r" );
	fprintf( stderr,
		"This program is distributed in the hope that it will be useful,\n\r"
		"but WITHOUT ANY WARRANTY; without even the implied warranty of\n\r"
		"MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n\r"
		"GNU General Public License for more details.\n\n\r" );

	fprintf( stderr, "Usage: gcmorse [OPTIONS]\n\r" );
	fprintf( stderr, "\t -a - append data to file instead of overwriting it\n\r" );
	fprintf( stderr, "\t -d - decrypt (from Morse to text)\n\r" );
	fprintf( stderr, "\t -h - show this help message\n\r" );
	fprintf( stderr, "\t -i <filename> - input file name (if not specified reads from stdin)\n\r" );
	fprintf( stderr, "\t -o <filename> - output file name (if not specified writes to stdout)\n\r" );
	fprintf( stderr, "\t -p - disable automatic prosign encryption\n\r" );
	fprintf( stderr, "\t -u - output uppercase text when decrypting\n\r" );
	fprintf( stderr, "\t -v - show version number\n\r" );
	fprintf( stderr, "\nSee also: `man gcmorse`\n\r" );
	exit( exitcode );
}

//Return pointer to an allocated part of memory containing string read from f
char *afgets( FILE *f )
{
	const size_t step = 1; //Amount of characters, that will be appended to string, each time we run out of space

	size_t i = 0, len = 8; //Current character index and initial str length
	char *str; //String
	int c; //Current character (int is better than char, ya know...)

	//Allocate space for first few characters
	if ( ( str = (char *) malloc( len ) ) == NULL ) return NULL;
	do //Repeat until c is 0, or realloc error occured
	{
		c = getc( f ); //Read single character
		//If current character pointer reaches end of string, allocate more memory
		if ( i >= len )
			if ( ( str = (char *) realloc( str, ( len += step ) ) ) == NULL ) return NULL;
		//Is character EOT or EOF? If so, set it to 0 to exit loop
		if ( c == 4 || c == EOF )
			c = 0;
		str[i++] = c; //Write character to array, and increment i
	}
	while ( c != 0 );
	return str;
}

//Text -> Morse conversion
void encrypt( FILE *outputfile, char *str, size_t len )
{
	unsigned char j, badc;
	size_t i;

	//Read characters from input file
	for ( i = 0; i < len; i++ )
	{
		badc = 1;

		//Iterate through pseudo-hash, looking for matches
		for ( j = 0; j < SUPPORTED_CHARACTERS && badc; j++ )
		{
			if ( tolower( str[i] ) == morse[j][0][0] )
			{
				if ( ( str[i] == '\n' || str[i] == '\r' ) && flags & FLAG_NOPROSIGNS )
				{
					badc = 0;
					continue;
				}

				//If next character is space, do not add additional one
				fprintf( outputfile, "%s%s", morse[j][1], ( i + 1 < len ) ? ( ( str[i + 1] == ' ' ) ? "" : " " ) : "" );
				badc = 0;
			}
		}

		//If no matches were found, throw a warning
		if ( badc ) fprintf( stderr, "gcmorse: unsupported character (ASCII only) - %c (0x%x) - c%ld\n\r", str[i], str[i], (long) i );
	}
}

//Morse -> Text conversion
void decrypt( FILE *outputfile, char *str, size_t len )
{
	char *morsechar = (char *) malloc( 1 );
	unsigned char badc, j, spaces = 0;
	size_t i, charend = 0;

	for ( i = 0; i < len; i++ )
	{
		badc = 1;

		//Ignore newlines etc.
		if ( str[i] == '\n' || str[i] == '\r' || str[i] == '\t' )
			continue;

		//Is space?
		if ( str[i] == ' ' )
		{
			spaces++;
			continue;
		}
		else
		{
			for ( j = 0; j < ( spaces >> 1 ); j++ ) fprintf( outputfile, " " );
			spaces = 0;
		}

		//Look for next meaningful character
		charend = i;
		while ( str[charend] != ' ' && str[charend] != '\0' && str[charend] != '\n' && str[charend] != '\r' && str[charend] != '\t' && charend < len )
			charend++;

		//Copy morse code to temporary array
		morsechar = (char *) realloc( morsechar, charend - i + 1 );
		memcpy( morsechar, str + i, charend - i );
		morsechar[charend - i] = 0;
		for ( j = 0; j < SUPPORTED_CHARACTERS; j++ )
		{
			if ( !strcmp( morsechar, morse[j][1] ) )
			{
				if ( ( morse[j][0][0] == '\n' || morse[j][0][0] == '\r' ) && flags & FLAG_NOPROSIGNS )
				{
					badc = 0;
					continue;
				}

				fprintf( outputfile, "%c", ( flags & FLAG_UPPERCASE ) ? toupper( morse[j][0][0] ) : morse[j][0][0] );
				badc = 0;
			}
		}

		if ( badc ) fprintf( stderr, "gcmorse: unsupported Morse code - '%s' - c%ld\n\r", morsechar, (long) i );

		i = charend - 1;
	}

	free( morsechar );
}

int main( int argc, char **argv )
{
	unsigned char i;
	char *inputstr = NULL, *inputfilename = NULL, *outputfilename = NULL, argparsed = 0;
	FILE *inputfile = NULL, *outputfile = NULL;
	size_t inputstrlen;

	//Search argv for supported arguments
	for ( i = 1; i < argc; i++ )
	{
		//Argument is assument to be not parsed as default
		argparsed = 0;

		//Show help message
		if ( !strcmp( argv[i], "-h" ) || !strcmp( argv[i], "--help" ) )
		{
			argparsed = 1;
			help( 0 );
		}

		//Show version number
		if ( !strcmp( argv[i], "-v" ) || !strcmp( argv[i], "--version" ) )
		{
			argparsed = 1;
			version( 0 );
		}

		//Decrpyt morse message
		if ( !strcmp( argv[i], "-d" ) || !strcmp( argv[i], "--decrypt" ) )
		{
			argparsed = 1;
			flags |= FLAG_DECRYPT;
		}

		//Output uppercase letters
		if ( !strcmp( argv[i], "-u" ) || !strcmp( argv[i], "--uppercase" ) )
		{
			argparsed = 1;
			flags |= FLAG_UPPERCASE;
		}

		//Output uppercase letters
		if ( !strcmp( argv[i], "-p" ) || !strcmp( argv[i], "--prosignsdisabled" ) )
		{
			argparsed = 1;
			flags |= FLAG_NOPROSIGNS;
		}

		//Append data
		if ( !strcmp( argv[i], "-a" ) || !strcmp( argv[i], "--append" ) )
		{
			argparsed = 1;
			flags |= FLAG_APPEND;
		}

		//Specify input file
		if ( !strcmp( argv[i], "-i" ) || !strcmp( argv[i], "--input" ) )
		{
			argparsed = 1;
			if ( i + 1 < argc )
			{
				inputfilename = argv[++i];
			}
			else
			{
				fprintf( stderr, "gcmorse: missing input file name.\nTry -h option to get more information.\n\r" );
				exit( 1 );
			}
		}

		//Specify output file
		if ( !strcmp( argv[i], "-o" ) || !strcmp( argv[i], "--output" ) )
		{
			argparsed = 1;
			if ( i + 1 < argc )
				outputfilename = argv[++i];
			else
			{
				fprintf( stderr, "gcmorse: missing output file name.\nTry -h option to get more information.\n\r" );
				exit( 1 );
			}
		}

		//Throw error if argument is unexpected
		if ( !argparsed )
		{
			fprintf( stderr, "gcmorse: unknown option: %s\n\r", argv[i] );
			help( 1 );
		}
	}

	//Open input file
	if ( inputfilename == NULL ) inputfile = stdin;
	else if ( ( inputfile = fopen( inputfilename, "r" ) ) == NULL  )
	{
		fprintf( stderr, "gcmorse: unable to open input file.\nTry -h option to get more information.\n\r" );
		exit( 1 );
	}

	DEBUG( "%d", inputfilename );
	inputstr = afgets( inputfile );
	inputstrlen = strlen( inputstr );

	//Close input file
	if ( inputfile != stdin ) fclose( inputfile );
	else if ( outputfilename == NULL ) printf( "\n" );

	//Open outputfile
	if ( outputfilename == NULL ) outputfile = stdout;
	else if ( ( outputfile = fopen( outputfilename, ( flags & FLAG_APPEND ) ? "a" : "w" ) ) == NULL )
	{
		fprintf( stderr, "gcmorse: unable to open output file.\nTry -h option to get more information.\n\r" );
		exit( 1 );
	}

	//Encrypt or decrypt file
	if ( flags & FLAG_DECRYPT )
		decrypt( outputfile, inputstr, inputstrlen );
	else
		encrypt( outputfile, inputstr, inputstrlen );

	//Print empty line or close input file
	if ( outputfile == stdout ) printf( "\n\r" );
	else fclose( outputfile );

	free( inputstr );

	return 0;
}
